### PR TITLE
feat: adding a new matcher for shadow DOM toMatch

### PR DESCRIPTION
**Pull request by k-j-kim**
## Summary

- based on @nolanlawson's `text-content-shadow-dom`, adding a new `toMatch`-like matcher that traverses shadow DOM
   - `shadowroot` must be set to `open` to allow traversal
   - should fix https://github.com/smooth-code/jest-puppeteer/issues/328

## Test plan

- `toMatchInShadow.test.js` that executes the same test cases as `toMatch`

### CODE REVIEW

**Comment by k-j-kim** [comment]
Since I'm unsure whether it's best to create a separate matcher for this or to augment `toMatch`, I'm creating this PR as a draft to gather feedback.

File: packages/expect-puppeteer/src/matchers/toMatchInShadow.js [hunk 1]
```diff
@@ -0,0 +1,78 @@
+import { getContext, enhanceError, expandSearchExpr } from '../utils'
+import { defaultOptions } from '../options'
+
+async function toMatch(instance, matcher, options) {
```
Comment (UziTech) [inline packages/expect-puppeteer/src/matchers/toMatchInShadow.js:4]:
```suggestion
async function toMatchInShadow(instance, matcher, options) {
```

File: packages/expect-puppeteer/src/matchers/toMatchInShadow.js [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+      handle,
+      text,
+      regexp,
+    )
+  } catch (error) {
+    throw enhanceError(error, `Text not found "${matcher}"`)
+  }
+}
+
+export default toMatch
```
Comment (UziTech) [inline packages/expect-puppeteer/src/matchers/toMatchInShadow.js:78]:
```suggestion
export default toMatchInShadow
```

**Comment by UziTech** [comment]
It looks like it would be better to add an option to `toMatch` since this doesn't only match if the text is in shadow dom.

**Comment by k-j-kim** [comment]
fair point @UziTech, let me update the PR.

File: packages/expect-puppeteer/src/matchers/notToMatch.js [hunk 3]
```diff
@@ -3,18 +3,63 @@ import { defaultOptions } from '../options'
...
+          }
+          return result
+        }
+
         if (!handle) return false
-        return handle.textContent.match(new RegExp(matcher)) === null
+
+        const textContent = traverseShadowRoots
+          ? getShadowTextContent(handle)
+          : handle.textContent
```
Comment (k-j-kim) [inline packages/expect-puppeteer/src/matchers/notToMatch.js:55]: get `textContent` from either `getShadowTextContent` or node's `textContent` based on `traverseShadowRoots` option.

File: packages/expect-puppeteer/src/matchers/notToMatch.test.js [hunk 4]
```diff
@@ -5,43 +5,51 @@ describe('not.toMatch', () => {
...
-
-      try {
-        await expect(page).not.toMatch('home')
-      } catch (error) {
-        expect(error.message).toMatch('Text found "home"')
-        expect(error.message).toMatch('waiting for function failed')
-      }
-    })
-  })
+  describe.each(['Page', 'Frame', 'ShadowPage', 'ShadowFrame'])(
```
Comment (k-j-kim) [inline packages/expect-puppeteer/src/matchers/notToMatch.test.js:8]: same test cases, just adding page types.

File: packages/expect-puppeteer/src/matchers/notToMatch.test.js [hunk 5]
```diff
@@ -5,43 +5,51 @@ describe('not.toMatch', () => {
...
-        await expect(dialogBtn).not.toMatch('Open dialog')
-      } catch (error) {
-        expect(error.message).toMatch('Text found "Open dialog"')
-        expect(error.message).toMatch('waiting for function failed')
-      }
-    })
-  })
+      const options = ['ShadowPage', 'ShadowFrame'].includes(pageType)
+        ? { traverseShadowRoots: true }
+        : {}
```
Comment (k-j-kim) [inline packages/expect-puppeteer/src/matchers/notToMatch.test.js:18]: run `toMatch` with option if in shadow DOM page type

File: packages/expect-puppeteer/src/matchers/notToMatch.test.js [hunk 6]
```diff
@@ -5,43 +5,51 @@ describe('not.toMatch', () => {
...
+
+        try {
+          await expect(page).not.toMatch('home', options)
+        } catch (error) {
+          expect(error.message).toMatch('Text found "home"')
+          expect(error.message).toMatch('waiting for function failed')
+        }
+      })
+
+      describe('ElementHandle', () => {
```
Comment (k-j-kim) [inline packages/expect-puppeteer/src/matchers/notToMatch.test.js:35]: moved `ElementHandle` into the page type block. Was there a reason it was not tested for frames?

**Review by UziTech — APPROVED** [review]
LGTM :100:

**Comment by k-j-kim** [comment]
tyvm for the review @UziTech! 🙏 

is there anything else I need to do to get the PR ready for merge?

**Comment by UziTech** [comment]
Nope. I will release this tonight. Thanks!

**Comment by UziTech** [comment]
Sorry for the delay. Finally got around to releasing 6.1.0 with this change 🎉

**Pull request merged**